### PR TITLE
PIM-3942: extract getExpectedValues from ProductBuilder

### PR DESCRIPTION
**Pull request by nidup**
| Q                    | A
| -------------------- | ---
| Bug fix?             | N
| New feature?         | N
| BC breaks?           | Y
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | Y
| Fixed tickets        | PIM-3942
| DB schema updated?   |
| Migration script?    |
| Doc PR               |

### CODE REVIEW

File: spec/Pim/Bundle/CatalogBundle/Builder/ProductBuilderSpec.php [hunk 1]
```diff
@@ -31,11 +32,10 @@ class ProductBuilderSpec extends ObjectBehavior
     function let(
         AttributeRepositoryInterface $attributeRepository,
         FamilyRepositoryInterface $familyRepository,
-        ChannelRepositoryInterface $channelRepository,
-        LocaleRepositoryInterface $localeRepository,
```
Comment (skeleton) [inline spec/Pim/Bundle/CatalogBundle/Builder/ProductBuilderSpec.php]: These namespaces are always declared on top of the file

Comment (nidup) [inline spec/Pim/Bundle/CatalogBundle/Builder/ProductBuilderSpec.php]: fixed, thx

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+     * Returns an array of values that are expected to link product to an attribute depending on locale and scope
+     * Each value is returned as an array with 'scope' and 'locale' keys
+     *
+     * @param AttributeInterface[] $attributes
+     *
+     * @return array:array
+     */
+    public function getEligibleValues(array $attributes)
+    {
+        $values = array();
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: use []

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        foreach ($attributes as $attribute) {
+            $requiredValues = array();
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: idem

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            if ($attribute->isScopable() && $attribute->isLocalizable()) {
+                $requiredValues = $this->getScopeToLocaleRows($attribute);
+            } elseif ($attribute->isScopable()) {
+                $requiredValues = $this->getScopeRows($attribute);
+            } elseif ($attribute->isLocalizable()) {
+                $requiredValues = $this->getLocaleRows($attribute);
+            } else {
+                $requiredValues[] = array('attribute' => $attribute->getCode(), 'locale' => null, 'scope' => null);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: idem

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+use Pim\Bundle\CatalogBundle\Repository\LocaleRepositoryInterface;
+
+/**
+ * Resolve expected values for attributes
+ *
+ * @author    Nicolas Dupont <nicolas@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class AttributesValuesResolver
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: AttributeValuesResolver ?

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+use Pim\Bundle\CatalogBundle\Model\AttributeInterface;
+use Pim\Bundle\CatalogBundle\Repository\ChannelRepositoryInterface;
+use Pim\Bundle\CatalogBundle\Repository\LocaleRepositoryInterface;
+
+/**
+ * Resolve expected values for attributes
+ *
+ * @author    Nicolas Dupont <nicolas@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: 2015 :)

File: src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    protected function getScopeToLocaleRows(AttributeInterface $attribute)
+    {
+        $channels = $this->channelRepository->findAll();
+        $scopeToLocaleRows = [];
+        foreach ($channels as $channel) {
+            foreach ($channel->getLocales() as $locale) {
+                $scopeToLocaleRows[] = [
+                    'attribute' => $attribute->getCode(),
+                    'locale' => $locale->getCode(),
+                    'scope' => $channel->getCode()
```
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributesValuesResolver.php]: pls align params

File: src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+     *
+     * @param AttributeInterface $attribute
+     * @param array              $values
+     *
+     * @return array
+     */
+    protected function filterExpectedValues(AttributeInterface $attribute, array $values)
+    {
+        if ($attribute->isLocaleSpecific()) {
+            $availableLocales = $attribute->getLocaleSpecificCodes();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php]: use an instance var  for this, you load them for each attr

File: src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Return rows for available locales
+     *
+     * @param AttributeInterface $attribute
+     *
+     * @return array
+     */
+    protected function getLocaleRows(AttributeInterface $attribute)
+    {
+        $locales = $this->localeRepository->getActivatedLocales();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php]: same here

File: src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Return rows for available channels and theirs locales
+     *
+     * @param AttributeInterface $attribute
+     *
+     * @return array
+     */
+    protected function getScopeToLocaleRows(AttributeInterface $attribute)
+    {
+        $channels = $this->channelRepository->findAll();
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php]: same here

File: src/Pim/Bundle/CatalogBundle/Builder/ProductBuilder.php [hunk 11]
```diff
@@ -162,7 +155,7 @@ public function addMissingAssociations(ProductInterface $product)
      */
     public function addAttributeToProduct(ProductInterface $product, AttributeInterface $attribute)
     {
-        $requiredValues = $this->getExpectedValues(array($attribute));
+        $requiredValues = $this->valuesResolver->getEligibleValues(array($attribute));
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Builder/ProductBuilder.php]: use brackets

File: src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+    /**
+     * Returns an array of values that are expected to link product to an attribute depending on locale and scope
+     * Each value is returned as an array with 'scope' and 'locale' keys
+     *
+     * @param AttributeInterface[] $attributes
+     *
+     * @return array:array
+     */
+    public function getEligibleValues(array $attributes)
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php]: a resolver usually resolves
either the method or the class is not named well enough, up to you

File: src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php (hunk 8 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Manager/AttributeValuesResolver.php]: :+1:

File: src/Pim/Bundle/CatalogBundle/Builder/ProductBuilder.php (hunk 11 above)
Comment (nidup) [inline src/Pim/Bundle/CatalogBundle/Builder/ProductBuilder.php]: thx i fix this in coming PR

**Pull request merged**